### PR TITLE
ARROW-13277: [JS] Add declaration maps for TypeScript and refactor testing infrastructure

### PR DESCRIPTION
**Pull request by domoritz**
With declaration maps, developers can jump to the implementation rather than a declaration file.

### CODE REVIEW

**Review by trxcllnt — APPROVED** [review]
:heart_eyes: I _love_ declaration maps :tada:

File: js/jest.config.js [hunk 1]
```diff
@@ -45,7 +45,8 @@ module.exports = {
       "^.+\\.tsx?$": "ts-jest"
     },
     "transformIgnorePatterns": [
-      "/node_modules/(?!web-stream-tools).+\\.js$"
+      "/node_modules/(?!web-stream-tools).+\\.js$",
+      ".*/cjs/.*\\.js$"
```
Comment (domoritz) [inline js/jest.config.js:49]: Without this, I am getting this strange error when I run `yarn gulp test:es5:cjs`?

```
ts-jest[ts-compiler] (WARN) Unable to process '/Users/dominik/Code/arrow/js/targets/es5/cjs/io/adapters.js', falling back to original file content. You can also configure Jest config option `transformIgnorePatterns` to ignore /Users/dominik/Code/arrow/js/targets/es5/cjs/io/adapters.js from transformation or make sure that `outDir` in your tsconfig is neither `''` or `'.'`
```

Any idea what's causing it?

Comment (domoritz) [reply js/jest.config.js:49]: It looks like our setup is unhappy about having ts files right next to js files. I think I can redo this on top of #10371.

**Comment by domoritz** [comment]
Hmm, can it be that test times for the docker build went from 20 minutes to 50 minutes? (https://github.com/apache/arrow/runs/3005016709 vs https://github.com/apache/arrow/runs/3037021755).

**Review by domoritz — COMMENTED** [review]
LGTM

**Review by trxcllnt — APPROVED** [review]

**Pull request closed without merging**